Subscription manager breaks our spawners

Original issue: https://pagure.io/copr/copr/issue/2132
Opened: 2022-03-22 16:08:23
Opened by: [praiskup](https://accounts.fedoraproject.org/user/praiskup)

It looks this happens in all the clouds now, but not 100% (occasionally some machine starts, and is successful with the subscription):

```

2022-03-22 16:08:00,532 [ERROR] subscription-manager:2142:MainThread @cli.py:61 - Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
Traceback (most recent call last):
  File "/usr/lib64/python3.10/site-packages/subscription_manager/cli_command/register.py", line 222, in _do_command
    owner_key = service.determine_owner_key(
  File "/usr/lib64/python3.10/site-packages/rhsmlib/services/register.py", line 180, in determine_owner_key
    owners = self.cp.getOwnerList(username)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 1383, in getOwnerList
    owners = self.conn.request_get(method)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 920, in request_get
    return self._request("GET", method, headers=headers, cert_key_pairs=cert_key_pairs)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 945, in _request
    result = super(Restlib, self)._request(request_type, method,
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 805, in _request
    self.validateResponse(result, request_type, handler)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 883, in validateResponse
    raise RemoteServerException(response['status'],
rhsm.connection.RemoteServerException: Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
2022-03-22 16:08:06,492 [ERROR] subscription-manager:2144:MainThread @connection.py:833 - Response: 502
2022-03-22 16:08:06,492 [ERROR] subscription-manager:2144:MainThread @connection.py:834 - JSON parsing error: Expecting value: line 1 column 1 (char 0)
2022-03-22 16:08:06,492 [ERROR] subscription-manager:2144:MainThread @cli.py:60 - Error during registration: Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
2022-03-22 16:08:06,492 [ERROR] subscription-manager:2144:MainThread @cli.py:61 - Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
Traceback (most recent call last):
  File "/usr/lib64/python3.10/site-packages/subscription_manager/cli_command/register.py", line 222, in _do_command
    owner_key = service.determine_owner_key(
  File "/usr/lib64/python3.10/site-packages/rhsmlib/services/register.py", line 180, in determine_owner_key
    owners = self.cp.getOwnerList(username)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 1383, in getOwnerList
    owners = self.conn.request_get(method)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 920, in request_get
    return self._request("GET", method, headers=headers, cert_key_pairs=cert_key_pairs)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 945, in _request
    result = super(Restlib, self)._request(request_type, method,
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 805, in _request
    self.validateResponse(result, request_type, handler)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 883, in validateResponse
    raise RemoteServerException(response['status'],
rhsm.connection.RemoteServerException: Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
2022-03-22 16:08:12,462 [ERROR] subscription-manager:2146:MainThread @connection.py:833 - Response: 502
2022-03-22 16:08:12,463 [ERROR] subscription-manager:2146:MainThread @connection.py:834 - JSON parsing error: Expecting value: line 1 column 1 (char 0)
2022-03-22 16:08:12,463 [ERROR] subscription-manager:2146:MainThread @cli.py:60 - Error during registration: Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
2022-03-22 16:08:12,463 [ERROR] subscription-manager:2146:MainThread @cli.py:61 - Server error attempting a GET to /subscription/users/copr-team/owners returned status 502
Traceback (most recent call last):
  File "/usr/lib64/python3.10/site-packages/subscription_manager/cli_command/register.py", line 222, in _do_command
    owner_key = service.determine_owner_key(
  File "/usr/lib64/python3.10/site-packages/rhsmlib/services/register.py", line 180, in determine_owner_key
    owners = self.cp.getOwnerList(username)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 1383, in getOwnerList
    owners = self.conn.request_get(method)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 920, in request_get
    return self._request("GET", method, headers=headers, cert_key_pairs=cert_key_pairs)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 945, in _request
    result = super(Restlib, self)._request(request_type, method,
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 805, in _request
    self.validateResponse(result, request_type, handler)
  File "/usr/lib64/python3.10/site-packages/rhsm/connection.py", line 883, in validateResponse
    raise RemoteServerException(response['status'],
rhsm.connection.RemoteServerException: Server error attempting a GET to /subscription/users/copr-team/owners returned status 502

```

---

#### [praiskup](https://accounts.fedoraproject.org/user/praiskup) commented at 2022-03-22 17:16:05:
I filled a Red Hat support case 03178902 for this.

---

#### [praiskup](https://accounts.fedoraproject.org/user/praiskup) commented at 2022-03-22 17:46:26:
This happened for the second time, and it blocked all users for a long time while only a subset of users depends on it.  We should rework it:
1. the subscription management task should be asynchronous from the spawning playbook, it takes an unreasonable amount of time anyway.  @msuchy did the discussion with S-M folks went somewhere?
2. only EL8 builds should be "blocked" ... and they could wait till the async process finishes?
3. perhaps fail the whole build if it 2. takes too much time?  Or try 3 times, on at most three different builders?

--- a/backend/copr_backend/background_worker_build.py
+++ b/backend/copr_backend/background_worker_build.py
@@ -242,21 +242,25 @@ class BuildBackgroundWorker(BackendBackgroundWorker):
             raise BuildRetry("Minimum version for builder is {}"
                              .format(MIN_BUILDER_VERSION))
 
-    def _check_mock_config(self):
-        config = "/etc/mock/{}.cfg".format(self.job.chroot)
-        command = "/usr/bin/test -f " + config
-        if self.job.chroot == "srpm-builds":
-            return
-        if self.ssh.run(command):
-            raise BuildRetry("Chroot config {} not found".format(config))
-
     def _check_vm(self):
         """
         Check that the VM is OK to start the build
         """
         self.log.info("Checking that builder machine is OK")
         self._check_copr_builder()
-        self._check_mock_config()
+
+        # The output won't be live and will appear only after this command
+        # finishes. Making it live is nontrivial but we have a good code for
+        # doing so in `resallocserver.manager.run_command`. Praiskup plans to
+        # generalize it into a separate package that we could eventually use
+        # here.
+        cmd = "copr-builder-ready " + self.job.chroot
+        rc, stdout, stderr = self.ssh.run_expensive(
+            cmd, subprocess_timeout=660)
+        self.log.info(stdout)
+        if rc:
+            self.log.info(stderr)
+            raise BuildRetry("Builder wasn't ready, trying a new one")
 
     def _fill_build_info_file(self):
         """

--- a/backend/tests/test_background_worker_build.py
+++ b/backend/tests/test_background_worker_build.py
@@ -695,12 +695,12 @@ def test_ssh_connection_error(f_build_rpm_case, caplog):
         def __call__(self):
             self.counter += 1
             if self.counter == 1:
-                return (1, "err stdout", "err stderr")
+                return (1, b"err stdout", "err stderr")
             return (0, "", "")
 
     config = f_build_rpm_case
     ssh = config.ssh
-    ssh.set_command("/usr/bin/test -f /etc/mock/fedora-30-x86_64.cfg",
+    ssh.set_command("copr-builder-ready fedora-30-x86_64",
                     0, "", "", return_action=_SideEffect())
     worker = config.bw
     worker.process()
@@ -859,7 +859,7 @@ def test_failed_build_retry(f_build_rpm_case, caplog):
         hosts[index].hostname = "1.2.3." + str(index)
     rhf.return_value.get_host.side_effect = hosts
     ssh = config.ssh
-    ssh.set_command("/usr/bin/test -f /etc/mock/fedora-30-x86_64.cfg",
+    ssh.set_command("copr-builder-ready fedora-30-x86_64",
                     1, "", "not found")
 
     config.bw.process()

--- a/backend/tests/testlib/__init__.py
+++ b/backend/tests/testlib/__init__.py
@@ -130,8 +130,8 @@ class FakeSSHConnection(SSHConnection):
         self.commands = {}
         self.set_command(COMMANDS["rpm_q_builder"],
                          0, "666\n", "")
-        self.set_command("/usr/bin/test -f /etc/mock/fedora-30-x86_64.cfg",
-                         0, "", "")
+        self.set_command("copr-builder-ready fedora-30-x86_64", 0, "", "")
+        self.set_command("copr-builder-ready srpm-builds", 0, "", "")
         self.set_command("copr-rpmbuild-log",
                          0, "build log stdout\n", "build log stderr\n")
         self.resultdir = "fedora-30-x86_64/00848963-example"

--- a/rpmbuild/bin/copr-builder-ready
+++ b/rpmbuild/bin/copr-builder-ready
@@ -1,0 +1,94 @@
+#! /usr/bin/python3
+
+"""
+Final checks that the builder machine is ready to be used
+
+Everything printed to STDOUT will be redirected to the copr-backend logs,
+STDERR will be ignored.
+"""
+
+import os
+import sys
+import time
+from fnmatch import fnmatch
+from copr_rpmbuild.config import Config
+
+
+def check_mock_config(chroot):
+    """
+    Does the mock config for this chroot exist?
+    """
+    if chroot == "srpm-builds":
+        return
+
+    config = "/etc/mock/{}.cfg".format(chroot)
+    if os.path.isfile(config):
+        return
+
+    print("Chroot config {} not found".format(config))
+    sys.exit(1)
+
+
+def subscription_required(chroot):
+    """
+    Is subscription required for this task?
+    """
+    config = Config()
+    config.load_config()
+
+    for pattern in config.rhsm:
+        if fnmatch(chroot, pattern):
+            return True
+    return False
+
+
+def active_subscription():
+    """
+    Is subscription active on this system?
+    """
+    # There are standard-ish ways for checking whether the subscription is
+    # active. No silver bullet, all of them have trade-offs.
+    #   - Checking the existence of `/etc/pki/consumer/cert.pem` file
+    #   - Checking the existence of `/etc/pki/entitlement/*.pem` files
+    #   - Exit code from `subscription-manager status`
+    #   - Exit code from `subscription-manager identity`
+    # We don't want to rely on any of them. We use a custom daemon for
+    # registering the system on the background. Once it is done,
+    # it creates a file.
+    return os.path.exists("/run/copr-builder/rhsm-subscribed")
+
+
+def wait_for_subscription(timeout=600):
+    """
+    Wait until this system has an active subscription
+
+    Activating Red Hat subscription may take a lot of time and historically, the
+    subscription service used to be unreliable, so we should wait for the
+    subscription only when necessary.
+    """
+    start = time.time()
+    attempt = 1
+    while True:
+        print("Checking Red Hat subscription (attempt #{0})".format(attempt))
+        if active_subscription():
+            print("Red Hat subscription active")
+            return
+        if time.time() > start + timeout:
+            print("Waiting for Red Hat subscription timeouted!")
+            sys.exit(1)
+        time.sleep(30)
+        attempt += 1
+
+
+def main():
+    """
+    The entrypoint for this script
+    """
+    chroot = sys.argv[1]
+    check_mock_config(chroot)
+    if subscription_required(chroot):
+        wait_for_subscription()
+
+
+if __name__ == "__main__":
+    main()

--- a/rpmbuild/copr-rpmbuild.spec
+++ b/rpmbuild/copr-rpmbuild.spec
@@ -286,6 +286,7 @@ EOF
 %_bindir/copr-builder-cleanup
 %_bindir/copr-builder-rhsm-subscribe
 %_bindir/copr-builder-rhsm-subscribe-daemon
+%_bindir/copr-builder-ready
 %_sysconfdir/copr-builder
 %dir %mock_config_overrides
 %doc %mock_config_overrides/README

--- a/rpmbuild/copr-rpmbuild.yml
+++ b/rpmbuild/copr-rpmbuild.yml
@@ -12,3 +12,8 @@
 #       cute
 #       multiline
 #       snippet
+#
+# Chroots that require active Red Hat subscription
+# rhsm:
+#   - rhel-*
+#   - epel-*

--- a/rpmbuild/copr_rpmbuild/config.py
+++ b/rpmbuild/copr_rpmbuild/config.py
@@ -12,8 +12,10 @@ class Config:
     """
     Configuration class for copr-rpmbuild
     """
+
     def __init__(self):
         self.tags_to_mock_snippet = []
+        self.rhsm = []
 
     def load_config(self):
         """
@@ -27,3 +29,4 @@ class Config:
             pass
 
         self.tags_to_mock_snippet = config_data.get("tags_to_mock_snippet", [])
+        self.rhsm = config_data.get("rhsm", [])

